### DISCUSSION TITLE
Type inference in extension methods

### DISCUSSION THREAD

**Discussion by alex-bello** [dotnet/csharplang#6275] [Q&A]
Is it possible to infer some of the type parameters in an extension method based on the type being extended? For example, let's say I have a builder class that has three types param with constraints.

```c#
public class BuilderClass<TUser, TRole, TUserRole>
  where TUser : AppUser, new()
  where TRole : AppRole, new()
  where TUserRole : AppUserRole<TUser, TRole>, new()
{
  ...
}

```

I'd like to add an extension method that has a type parameter with a constraint as well that also relies on other type parameters with constraints. However, I'd like to avoid having to list the same type parameters over and over with each extension method that has a new type parameter.

```c#
public static class ExtensionMethods
{
  public static BuilderClass<TUser, TRole, TUserRole> UseSqlDb<TDbContext, TUser, TRole, TUserRole>
  ( 
    this BuilderClass<TUser, TRole, TUserRole> builder,
    string connectionString
  )
    where TDbContext : AppDbContext<TUser, TRole, TUserRole>
    where TUser : AppUser, new()
    where TRole : AppRole, new()
    where TUserRole : AppUserRole<TUser, TRole>, new()
  {
    builder.AddDbContext<TDbContext>(connectionString);
  }
}
```

I was hoping that I'd be able to just do the following:

```c#
var appBuilder = new BuilderClass<UserType, RoleType, UserRoleType>();

builder.UseSqlDb<SqlDbContext>("[ConnectionString]");
```

However, the compiler does not recognize the extension method. Is there a way to make this possible?

**Comment by svick** [comment]
I don't think extension methods are or should be special here. Which means this should work if both https://github.com/dotnet/csharplang/issues/1349 and https://github.com/dotnet/csharplang/discussions/741 were implemented.
